geode-core/src/main/java/org/apache/geode/internal/net/BufferPool.java

```diff
@@ -67,39 +68,18 @@ public ByteBuffer acquireReceiveBuffer(int size) {
   private ByteBuffer acquireBuffer(int size, boolean send) {
     ByteBuffer result;
     if (useDirectBuffers) {
-      IdentityHashMap<BBSoftReference, BBSoftReference> alreadySeen = null; // keys are used like a
-                                                                            // set
-      BBSoftReference ref = bufferQueue.poll();
-      while (ref != null) {
-        ByteBuffer bb = ref.getBB();
-        if (bb == null) {
-          // it was garbage collected
-          int refSize = ref.consumeSize();
-          if (refSize > 0) {
-            if (ref.getSend()) { // fix bug 46773
-              stats.incSenderBufferSize(-refSize, true);
-            } else {
-              stats.incReceiverBufferSize(-refSize, true);
-            }
-          }
-        } else if (bb.capacity() >= size) {
-          bb.rewind();
-          bb.limit(size);
-          return bb;
-        } else {
-          // wasn't big enough so put it back in the queue
-          Assert.assertTrue(bufferQueue.offer(ref));
-          if (alreadySeen == null) {
-            alreadySeen = new IdentityHashMap<>();
-          }
-          if (alreadySeen.put(ref, ref) != null) {
-            // if it returns non-null then we have already seen this item
-            // so we have worked all the way through the queue once.
-            // So it is time to give up and allocate a new buffer.
-            break;
-          }
+      for(ByteBufferSoftReference ref : bufferQueue) {
+        bufferQueue.remove(ref);
+
+        if(garbageCollected(ref)){
```

Comment:
This code has a subtle bug. You need the "garbageCollected" method to return a strong reference to the non-null ByteBuffer. If you don't then, after you call garbageCollected, it could still become null and lead to NPEs in the following code.

Reply:
👍 good catch, thanks